### PR TITLE
Add support for Unnest of Map column

### PR DESCRIPTION
**Pull request by aditi-pandit**

### CODE REVIEW

**Comment by aditi-pandit** [comment]
@mbasmanova : This is what I intended for Unnest with Map column. 

Unfortunately it seems like DuckDB doesn't yet support UNNEST of MAP column so all the wiring with DuckDB went to waste.  But I'll keep it around for any later use.

I'm still working on the tests though.

File: velox/exec/Unnest.h [hunk 1]
```diff
@@ -39,7 +39,10 @@ class Unnest : public Operator {
   bool isFinished() override;
 
  private:
+  enum UnnestColumnType { ARRAY, MAP };
```
Comment (mbasmanova) [inline velox/exec/Unnest.h:42]: use existing TypeKind, e.g. 

```
TypeKind unnestColumnType_;

unnestColumnType_ = unnestVariable->type()->kind();
```

File: velox/exec/Unnest.cpp [hunk 2]
```diff
@@ -64,18 +67,26 @@ RowVectorPtr Unnest::getOutput() {
...
   auto unnestIndices = unnestDecoded_.indices();
 
-  auto unnestBase = unnestDecoded_.base()->as<ArrayVector>();
-  auto rawSizes = unnestBase->rawSizes();
-  auto rawOffsets = unnestBase->rawOffsets();
+  const ArrayVector* unnestDecodedBaseArrayVector;
+  const MapVector* unnestDecodedBaseMapVector;
+  const vector_size_t* rawSizes;
+  const vector_size_t* rawOffsets;
+  if (unnestColumnType_ == ARRAY) {
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:80]: unnestColumnType_ is not needed

```
if (unnestVector->typeKind() == TypeKind::ARRAY) {
   ...
}
```

File: velox/exec/Unnest.cpp [hunk 3]
```diff
@@ -64,18 +67,26 @@ RowVectorPtr Unnest::getOutput() {
...
-
-  inputRows_.resize(size);
   unnestDecoded_.decode(*unnestVector, inputRows_);
-
   auto unnestIndices = unnestDecoded_.indices();
 
-  auto unnestBase = unnestDecoded_.base()->as<ArrayVector>();
-  auto rawSizes = unnestBase->rawSizes();
-  auto rawOffsets = unnestBase->rawOffsets();
+  const ArrayVector* unnestDecodedBaseArrayVector;
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:76]: These names are very long. Consider shortening to baseArray and baseMap or unnestArrayBase and unnestMapBase.

File: velox/exec/tests/UnnestTest.cpp [hunk 4]
```diff
@@ -39,6 +39,19 @@ TEST_F(UnnestTest, basic) {
   assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
 }
 
+TEST_F(UnnestTest, basicMap) {
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:42]: Consider creating DuckDB table with 3 columns, c0, c1.keys and c1.values, then unnesting 2 array columns in the query.

**Review by mbasmanova — COMMENTED** [review]
@aditi-pandit Aditi, looks nice. Some comments and suggestions below.

File: velox/exec/tests/UnnestTest.cpp (hunk 4 above)
Comment (aditi-pandit) [reply velox/exec/tests/UnnestTest.cpp:42]: Yeah, this approach worked. Thanks for the suggestion.

Have also added another test for empty and null maps.

**Comment by aditi-pandit** [comment]
> @aditi-pandit Looks good to me. The changes in QueryAssertions allow us to create DuckDB tables with MAP columns, right? These changes do not seem to be used in the diff. Do they work? It would be nicer to move these into a separate PR and add a test that leverages that functionality.

Yes, those changes were for supporting tables with MAP columns in DuckDB. I had started writing a test where I constructed such a table and then use it for UNNEST. The code I wrote was needed for the DuckDB table construction in createDuckDBTable(...) call. But then of course the UNNEST was not supported and assertQuery(...) failed.

I'll look at other tests and add something with a MAP column to keep this code exercised in a future PR. wdyt ?

**Comment by mbasmanova** [comment]
> I'll look at other tests and add something with a MAP column to keep this code exercised in a future PR. wdyt ?

Thank you for explaining. I'm going to remove changes to QueryAssertions from this diff and merge.

**Comment by aditi-pandit** [comment]
> > I'll look at other tests and add something with a MAP column to keep this code exercised in a future PR. wdyt ?
> 
> Thank you for explaining. I'm going to remove changes to QueryAssertions from this diff and merge.

That is fine also. Thanks.

**Comment by aditi-pandit** [comment]
#1066

**Pull request closed without merging**